typesetters/base.lua

```diff
@@ -728,42 +728,53 @@ function typesetter:breakpointsToLines (breakpoints)
         end
       end
       if seenHbox == 0 then break end
-      local mrg = self:getMargins()
-      self:addrlskip(slice, mrg, point.left, point.right)
-      local ratio = self:computeLineRatio(point.width, slice)
-      -- TODO see bug 620
-      -- if math.abs(ratio) > 1 then SU.warn("Using ratio larger than 1" .. ratio) end
-      local thisLine = { ratio = ratio, nodes = slice }
-      lines[#lines+1] = thisLine
+
+      -- If the line ends with a discretionary, repeat it on the next line,
+      -- so as to account for a potential postbreak.
       if slice[#slice].is_discretionary then
         linestart = point.position
       else
         linestart = point.position + 1
       end
+
+      -- Then only we can add some extra margin glue...
+      local mrg = self:getMargins()
+      self:addrlskip(slice, mrg, point.left, point.right)
+
+      -- And compute the line...
+      local ratio = self:computeLineRatio(point.width, slice)
```

Comment:
Erm. Could we do it earlier, before adding the margin skips (and related zerohbox'es etc.), since one of the things it does is actually to skip those? I was tempted initially (it simplifies its code), but it would then break compatibility. E.g. I used the method here, to recompute a full line:

https://github.com/Omikhleia/ptable.sile/blob/9b9ec3bbb5539acb8938fb094a50915facfb4f25/packages/parbox/init.lua#L161

Reply:
I also have several places where I'm calling that directly, possible for similar off-label use case.

Also I'm not 100% sure it would be *correct* to do it earlier before margins are added. For example I have a project where I'm using a hybrid justifications scheme using a small but *stretchable and shrinkable* right margen. It's not like a full ragged scenario where inter-word spaces are always the same size and the right margin is infinitely flexible, but it allows only relatively minor stretching inside the line that is offset with a little bit more generous stretch *at* the margin. This keeps hyphenation to a minimum even with chunky words in abundance (Turkish can be a bit of a mouth-full) which is especially desirable for the context while still allowing it to hyphenate when it really needs to without bad artifacts like overflows or half-full lines.

In that case not including the margins in the *ratio* of correction being applied to stretches would be a mistake.

Reply:
Agreed. So let's keep the current behavior and API for now. I do think at some point we may wish to revisit the typesetter (and the entanglement with frames, pagebuilder and insertions, notably...), but until then, fixing issues without breaking anything else is the way to go.